segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentMapper.java

```diff
@@ -515,9 +560,61 @@ private boolean isValidStreamSegmentId(long id) {
 
     //endregion
 
+    //region Helper Classes
+
     @Data
     private static class SegmentInfo {
         private final long segmentId;
         private final SegmentProperties properties;
     }
+
+    /**
+     * A pending request for a Segment Assignment, which keeps track of all queued callbacks.
+     * Note that this class in itself is not thread safe, so the caller should take precautions to ensure thread safety.
+     */
+    private static class PendingRequest {
+        private final ArrayList<QueuedCallback<?>> callbacks = new ArrayList<>();
+
+        /**
+         * Invokes all queued callbacks, in order, with the given SegmentId as a parameter.
+         */
+        void complete(long segmentId) {
+            for (QueuedCallback<?> callback : this.callbacks) {
+                try {
+                    callback.complete(segmentId);
```

Comment:
This is an inner class. It is at the mercy of its enclosing class. If you look at how it's used, you'll see that all access to its list of callbacks is done exclusively under the lock. Using a thread safe list won't help at all; it will just make things slower.